### PR TITLE
Inventory system

### PR DESCRIPTION
**Pull request by zshainsky**
## What does this PR do

1. Introduces some sprites we can use for testing
2. Creates Resources for various items and an inventory to use for testing
3. Creates ItemData, SlotData, and InventoryData classes to handle the resource's data functions
4. Creates Slot and Inventory scripts to connect the Inventory UI to the data resources
5. Adds Player Inventory UI to the main scene and introduces mouse click functionality for moving and stacking resources in the Inventory UI

## Resources

- Tutorial I followed: https://www.youtube.com/watch?v=V79YabQZC1s
- Sprites: https://cheekyinkling.itch.io/shikashis-fantasy-icons-pack

## Example/Controls
Controls

- Tab to toggle inventory UI
- Left mouse click to grab and stack. Stack limit size is x99 per slot if item is stackable
- Right mouse click to drop a single item from the grabbed item stack

Screen Shot 2023-04-16 at 12 54 42 PM

## TODO
Will continue messing around with different stuff...

1. Add Use/Equip functions for the player and specific Resource / Inventory types to support this
2. Be able to pickup items from the world and add to inventory
3. Maybe interact with other entities inventories

### CODE REVIEW

**Comment by zshainsky** [comment]
Dang these PR's can balloon to a lot of changed files fast 🙃

File: game/project.godot [hunk 1]
```diff
@@ -18,15 +18,23 @@ config/icon="res://icon.svg"
 [display]
 
 window/size/viewport_width=240
-window/size/viewport_height=240
+window/size/viewport_height=180
 window/size/window_width_override=1280
-window/size/window_height_override=1280
+window/size/window_height_override=1080
```
Comment (michaelwiest) [inline game/project.godot:23]: Any particular reason why we changed this?

File: game/project.godot [hunk 2]
```diff
@@ -18,15 +18,23 @@ config/icon="res://icon.svg"
...
 
 enabled=PackedStringArray("res://addons/gdfxr/plugin.cfg")
 
+[input]
+
+inventory={
+"deadzone": 0.5,
+"events": [Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":-1,"window_id":0,"alt_pressed":false,"shift_pressed":false,"ctrl_pressed":false,"meta_pressed":false,"pressed":false,"keycode":0,"physical_keycode":4194306,"key_label":0,"unicode":0,"echo":false,"script":null)
+]
+}
```
Comment (michaelwiest) [inline game/project.godot:36]: What is this doing? Or at least where in the GUI is this configured?

File: game/scripts/player/player.gd [hunk 3]
```diff
@@ -33,6 +37,9 @@ func _physics_process(delta):
 		var acc: float = (weapon_force / mass)
 		velocity += Vector2(last_direction.x * acc, last_direction.y * acc)
 	move_and_slide()
+	
+	# Open inventory menu if action pressed
+	inventory_input()
```
Comment (michaelwiest) [inline game/scripts/player/player.gd:42]: I think this is fine for now :+1 When we work on the player entity in earnest we'll probably want a more generalizable  way to look through inputs

File: game/scripts/player/player.gd [hunk 4]
```diff
@@ -43,3 +50,7 @@ func _on_stick_move_player(player_force):
 
 func _on_weapon_force_timer_timeout():
 	weapon_force = 0
+	
+func inventory_input() -> void:
+	if Input.is_action_just_pressed("inventory"):
```
Comment (michaelwiest) [inline game/scripts/player/player.gd:55]: Is this set in the `input_map`?

File: game/scripts/World.gd [hunk 5]
```diff
@@ -1,13 +1,28 @@
...
+	inventory_interface.set_player_inventory_data(player.inventory_data)
+	player.toggle_inventory.connect(toggle_inventory_interface)
 
 
 # Hack for now before having equipment associated with the player
 func _process(delta):
 	if Input.is_action_just_pressed("ui_accept"):
 		stick.use0()
+
+func toggle_inventory_interface() -> void:
```
Comment (michaelwiest) [inline game/scripts/World.gd:18]: Also fine for now but it feels weird to me to have this interface be a part of the World node. Is that the typical design pattern? feels like having some inventoryHandler being a chid of the player makes more sense to me? IDK or a set of UI nodes?

Comment (michaelwiest) [reply game/scripts/World.gd:18]: ^ I say this having this shitty stick input code here too lol

File: game/scripts/inventory/inventory_data.gd [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+signal inventory_interact(inventory_data: InventoryData, node_index: int, button_index: int)
+
+@export var slot_datas: Array[SlotData]
+
+func on_slot_clicked(node_index: int, button_index: int) -> void:
+	inventory_interact.emit(self, node_index, button_index)
+
+func grab_slot_data(index: int) -> SlotData:
+	var slot_data = null
+	if index >= 0 and index < len(slot_datas):
```
Comment (michaelwiest) [inline game/scripts/inventory/inventory_data.gd]: nit: datas not really a word.

File: game/scripts/inventory/slot.gd [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+	tooltip_text = "%s\n$s" % [item_data.name, item_data.description]
+	
+	if slot_data.quantity > 1:
+		quantity_label.text = "x%s" % slot_data.quantity
+		quantity_label.visible = true
+	else:
+		quantity_label.hide()
+
+
+## Hnadle GUI events on the inventory slot
```
Comment (michaelwiest) [inline game/scripts/inventory/slot.gd]: nit: typo

File: game/scripts/inventory/slot_data.gd [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+@export_range(1, MAX_STACK_SIZE) var quantity: int = 1: set = set_quantity
+
+func set_quantity(value):
+	quantity = value
+	
+	if quantity > 1 and not item_data.stackable:
+		quantity = 1
+		push_error("%s is not stackable, setting quantity to 1" % item_data.name)
+	
+func can_merge_with(other_slot_data: SlotData) -> bool:
```
Comment (michaelwiest) [inline game/scripts/inventory/slot_data.gd:15]: Cool!

**Review by michaelwiest — APPROVED** [review]
Nice I have some nits and some questions but works as expected!

File: game/project.godot (hunk 1 above)
Comment (zshainsky) [reply game/project.godot:23]: This just scales the screen when you run the game

Comment (zshainsky) [reply game/project.godot:23]: I updated the window viewport size to 240x180 and then scaled it up to a larger window size. I didn't love the square window that we had...no other reason

File: game/project.godot (hunk 2 above)
Comment (zshainsky) [reply game/project.godot:36]: I think this the InputMap that I setup for toggling the inventory UI:
Screen Shot 2023-04-19 at 8 34 37 AM

File: game/scripts/player/player.gd (hunk 4 above)
Comment (zshainsky) [reply game/scripts/player/player.gd:55]: Ye

**Pull request merged**